src/zenml/integrations/kserve/services/kserve_deployment.py

```diff
@@ -0,0 +1,569 @@
+    V1beta1PredictorSpec,
+    constants,
+)
+from kubernetes import client as k8s_client
+from pydantic import Field, ValidationError
+
+from zenml import __version__
+from zenml.logger import get_logger
+from zenml.services import (
+    BaseService,
+    ServiceConfig,
+    ServiceState,
+    ServiceStatus,
+    ServiceType,
+)
+
+if TYPE_CHECKING:
+    from zenml.integrations.kserve.model_deployers.kserve_model_deployer import (  # noqa
+        KServeModelDeployer,
+    )
+
+logger = get_logger(__name__)
+
+
+class KServeDeploymentConfig(ServiceConfig):
+    """KServe deployment service configuration.
+
+    Attributes:
+        model_uri: URI of the model (or models) to serve.
+        model_name: the name of the model. Multiple versions of the same model
+            should use the same model name.
+        predictor: the KServe predictor used to serve the model.
+        replicas: number of replicas to use for the prediction service.
+        resources: the Kubernetes resources to allocate for the prediction service.
+    """
+
+    model_uri: str = ""
+    model_name: str = "default"
+    secret_name: Optional[str]
+    predictor: str
+    replicas: int = 1
+    resources: Optional[Dict[str, Any]]
+
+    @staticmethod
+    def sanitize_labels(labels: Dict[str, str]) -> None:
+        """Update the label values to be valid Kubernetes labels.
+
+        See: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+
+        Args:
+            labels: The labels to sanitize.
+        """
+        # TODO[MEDIUM]: Move k8s label sanitization to a common module with all K8s utils.
+        for key, value in labels.items():
+            # Kubernetes labels must be alphanumeric, no longer than
+            # 63 characters, and must begin and end with an alphanumeric
+            # character ([a-z0-9A-Z])
+            labels[key] = re.sub(r"[^0-9a-zA-Z-_\.]+", "_", value)[:63].strip(
+                "-_."
+            )
+
+    def get_kubernetes_labels(self) -> Dict[str, str]:
+        """Generate the labels for the KServe inference CRD from the service configuration.
+
+        These labels are attached to the KServe inference service CRD
+        and may be used as label selectors in lookup operations.
+
+        Returns:
+            The labels for the KServe inference service CRD.
+        """
+        labels = {"app": "zenml"}
+        if self.pipeline_name:
+            labels["zenml.pipeline_name"] = self.pipeline_name
+        if self.pipeline_run_id:
+            labels["zenml.pipeline_run_id"] = self.pipeline_run_id
+        if self.pipeline_step_name:
+            labels["zenml.pipeline_step_name"] = self.pipeline_step_name
+        if self.model_name:
+            labels["zenml.model_name"] = self.model_name
+        if self.model_uri:
+            labels["zenml.model_uri"] = self.model_uri
+        if self.predictor:
+            labels["zenml.model_type"] = self.predictor
```

Comment:
here we are essentially doing the same 6 times, feels like it could be cleaner as a loop.